tests/ignite/metrics/test_multilabel_confusion_matrix.py

```diff
@@ -0,0 +1,438 @@
+
+    with pytest.raises(
+        ValueError, match=r"y must be of any type: \(torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64\)"
+    ):
+        cm.update((torch.rand(10, 10).type(torch.int32), torch.rand(10, 10)))
+
+    with pytest.raises(ValueError, match=r"y_pred must be a binary tensor"):
+        y = torch.randint(0, 2, size=(10, 10)).long()
+        y_pred = torch.randint(0, 2, size=(10, 10)).long()
+        y_pred[0, 0] = 2
+        cm.update((y_pred, y))
+
+    with pytest.raises(ValueError, match=r"y must be a binary tensor"):
+        y = torch.randint(0, 2, size=(10, 10)).long()
+        y_pred = torch.randint(0, 2, size=(10, 10)).long()
+        y[0, 0] = 2
+        cm.update((y_pred, y))
+
+
+def get_y_true_y_pred():
+    # Generate an image with labels 0 (background), 1, 2
+    # 3 classes:
+    y_true = np.zeros((1, 3, 30, 30), dtype=np.int64)
+    y_true[0, 0, 5:17, 7:11] = 1
+    y_true[0, 1, 1:11, 1:11] = 1
+    y_true[0, 2, 15:25, 15:25] = 1
+
+    y_pred = np.zeros((1, 3, 30, 30), dtype=np.int64)
+    y_pred[0, 0, 0:7, 8:15] = 1
+    y_pred[0, 1, 5:15, 1:11] = 1
+    y_pred[0, 2, 20:30, 20:30] = 1
+    return y_true, y_pred
+
+
+def test_multiclass_images():
+    num_classes = 3
+    cm = MultiLabelConfusionMatrix(num_classes=num_classes)
+
+    y_true, y_pred = get_y_true_y_pred()
+
+    # Compute confusion matrix with sklearn
+    sklearn_CM = multilabel_confusion_matrix(
+        y_true.transpose((0, 2, 3, 1)).reshape(-1, 3), y_pred.transpose((0, 2, 3, 1)).reshape(-1, 3)
+    )
+
+    # Update metric
+    output = (torch.tensor(y_pred), torch.tensor(y_true))
+    cm.update(output)
+
+    ignite_CM = cm.compute().cpu().numpy()
+
+    assert np.all(ignite_CM == sklearn_CM)
+
+    # Another test on batch of 2 images
+    cm = MultiLabelConfusionMatrix(num_classes=num_classes)
+
+    # Create a batch of two images:
+    th_y_true1 = torch.tensor(y_true)
+    th_y_true2 = torch.tensor(y_true.transpose(0, 1, 3, 2))
+    th_y_true = torch.cat([th_y_true1, th_y_true2], dim=0)
+
+    th_y_pred1 = torch.tensor(y_pred)
+    th_y_pred2 = torch.tensor(y_pred.transpose(0, 1, 3, 2))
+    th_y_pred = torch.cat([th_y_pred1, th_y_pred2], dim=0)
+
+    # Update metric & compute
+    output = (th_y_pred, th_y_true)
+    cm.update(output)
+    ignite_CM = cm.compute().cpu().numpy()
+
+    # Compute confusion matrix with sklearn
+    th_y_true = idist.all_gather(th_y_true)
+    th_y_pred = idist.all_gather(th_y_pred)
+
+    np_y_true = th_y_true.cpu().numpy().transpose((0, 2, 3, 1)).reshape(-1, 3)
+    np_y_pred = th_y_pred.cpu().numpy().transpose((0, 2, 3, 1)).reshape(-1, 3)
+    sklearn_CM = multilabel_confusion_matrix(np_y_true, np_y_pred)
+
+    assert np.all(ignite_CM == sklearn_CM)
+
+
+def _test_distrib_multiclass_images(device):
+    def _test(metric_device):
+        num_classes = 3
+        cm = MultiLabelConfusionMatrix(num_classes=num_classes, device=metric_device)
+
+        y_true, y_pred = get_y_true_y_pred()
```

Comment:
We have to generate data a bit differently depending on distributed rank.

Reply:
Do you have any suggestions?

Reply:
Oh, I see, our tests for confusion matrix have the same remark. I think we should rework them. Let me propose a better tests for our confusion matrix which could be adapted for this PR as well. We can inspire from here : https://github.com/pytorch/ignite/blob/e17acc740d38909ddbbed6648423750abe593e58/tests/ignite/metrics/test_recall.py#L784-L785